File: samples/FsBulletML.Sample.MonoGame.FSharp/Player.fs
namespace FsBulletML.Sample.MonoGame.FSharp

open System
open System.Xml 
open System.Collections.Generic
open System.Runtime.Serialization
open Microsoft.Xna.Framework
open Microsoft.Xna.Framework.Input
open Microsoft.Xna.Framework.Graphics
open FsBulletML
open FsBulletML.MonoGame
 
type Player () as this =
  [<DefaultValue>]val mutable timer : int
  [<DefaultValue>]val mutable pos : Vector2
  [<DefaultValue>]val mutable speed : float32
  [<DefaultValue>]val mutable radius : float32
  [<DefaultValue>]val mutable damageCounter : int32
  [<DefaultValue>]val mutable texture : Texture2D

  let shoot2WayLeftBullet (player:Player) =
#if NET35 
    let ``2wayLeftBullet`` = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\2wayLeft.xml") |> BulletRunner.convertBulletmlTask |> Some
#endif
#if NET40 
    let ``2wayLeftBullet`` = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\2wayLeft.xml", None)  |> BulletRunner.convertBulletmlTask |> Some
#endif
    if player.timer > 0 then
      let bullet = new PlayerBullet()
      Manager.addPlayerBulletPos(bullet, new Vector2(this.pos.X - 10.f, this.pos.Y + 1.f))
      bullet.SetTask(``2wayLeftBullet``) 

  let shoot2WayRightBullet (player:Player) =
#if NET35 
    let ``2wayRightBullet`` = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\2wayRight.xml") |> BulletRunner.convertBulletmlTask |> Some
#endif
#if NET40 
    let ``2wayRightBullet`` = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\2wayRight.xml", None)  |> BulletRunner.convertBulletmlTask |> Some
#endif
    if player.timer > 0 then
      let bullet = new PlayerBullet()
      Manager.addPlayerBulletPos(bullet, new Vector2(this.pos.X + 10.f, this.pos.Y + 1.f ))
      bullet.SetTask(``2wayRightBullet``) 

  let shootHomingBullet (player:Player) = 
#if NET35 
    let homingBullet = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\homing.xml") |> BulletRunner.convertBulletmlTask |> Some
#endif
#if NET40
    let homingBullet = Bulletml.readXml (@"..\..\..\Content\xml\PlayerBullet\homing.xml", None) |> BulletRunner.convertBulletmlTask |> Some
#endif
    if player.timer > 60 then
      let bullet = new PlayerBullet()
      Manager.addPlayerBulletPos(bullet, this.pos)
      bullet.SetTask(homingBullet) 

  do
    this.pos <- new Vector2()
    this.speed <- 3.f
    this.radius <- 3.5f

  member this.Pos with get () = this.pos
                   and set (v) = this.pos <- v 

  member this.Speed with get () = this.speed
                     and set (v) = this.speed <- v 

  member this.Radius with get () = this.radius 
                      and set (v) = this.radius <- v
  member this.Init () =
    this.pos.X <- Settings.Player.X
    this.pos.Y <- Settings.Player.Y

  member this.Update () = 

    if (Keyboard.GetState().IsKeyDown(Keys.Left)) then
      if this.pos.X - this.speed  >= 0.f then
        this.pos.X <- this.pos.X - this.speed
    if (Keyboard.GetState().IsKeyDown(Keys.Right)) then
      if this.pos.X - this.speed <= Settings.Display.Width then
        this.pos.X <- this.pos.X +  this.speed
    if (Keyboard.GetState().IsKeyDown(Keys.Up)) then
      if this.pos.Y - this.speed >= 0.f then
        this.pos.Y <- this.pos.Y - this.speed
    if (Keyboard.GetState().IsKeyDown(Keys.Down)) then
      if this.pos.Y - this.speed <= Settings.Display.Height then
        this.pos.Y <- this.pos.Y + this.speed

    this.timer <- this.timer + 1
    if Keyboard.GetState().IsKeyDown(Keys.Z) then
      [shoot2WayLeftBullet; shoot2WayRightBullet; shootHomingBullet] |> List.iter (fun f -> f this)

    if (this.timer > 60) then
       this.timer <- 0

File: samples/FsBulletML.Sample.MonoGame.CSharp/Player.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

using FsBulletML;
using FsBulletML.MonoGame;
using Bulletml = FsBulletML.DTD.Bulletml;
using Settings = FsBulletML.MonoGame.Settings;

namespace FsBulletML.Sample.MonoGame.CSharp
{
    public class Player
    {
        private static Bulletml b2wayLeftBullet = Xml.readXml(@"..\..\Content\xml\PlayerBullet\2wayLeft.xml");
        private static Bulletml b2wayRightBullet = Xml.readXml(@"..\..\Content\xml\PlayerBullet\2wayRight.xml");
        private static Bulletml homing = Xml.readXml(@"..\..\Content\xml\PlayerBullet\homing.xml");

        private int Timer { get; set; }
        public Vector2 Pos { get; set; }
        public float Speed { get; set; }
        public float Radius { get; set; }
        public int DamageCounter { get; set; }
        public Texture2D Texture { get; set; }
        public Player() {}

        public void Init() 
        {
            this.Pos = new Vector2(Settings.Player.X, Settings.Player.Y);
            this.Speed = 3.5f;
            this.Radius = 3.5f;
        }

        public void Update()
        {
            var keyboardState = Keyboard.GetState();

            if (keyboardState.IsKeyDown(Keys.Left)) 
            {
                if (this.Pos.X - this.Speed  >= 0)
                    this.Pos = new Vector2(this.Pos.X - this.Speed, this.Pos.Y);
            }
            if (keyboardState.IsKeyDown(Keys.Right)) 
            {
                if (this.Pos.X - this.Speed <= Settings.Display.Width)
                    this.Pos = new Vector2(this.Pos.X + this.Speed, this.Pos.Y);
            }
            if (keyboardState.IsKeyDown(Keys.Up)) 
            {
                if (this.Pos.Y - this.Speed >= 0)
                    this.Pos = new Vector2(this.Pos.X, this.Pos.Y - this.Speed);
            }
            if (keyboardState.IsKeyDown(Keys.Down)) 
            {
                if (this.Pos.Y - this.Speed <= Settings.Display.Height)
                    this.Pos = new Vector2(this.Pos.X, this.Pos.Y + this.Speed);
            }

            this.Timer += 1;
            if (keyboardState.IsKeyDown(Keys.Z))
            {
                Shoot2WayLeftBullet();
                Shoot2WayRightBullet();
                ShootHomingBullet();
            }
 
            if (this.Timer > 60)
                this.Timer = 0;
        }

        private void Shoot2WayLeftBullet () 
        {
            var task = BulletRunner.ConvertBulletmlTaskOption(Player.b2wayLeftBullet);

            if (this.Timer > 0)
            {
                var bullet = new PlayerBullet();
                Manager.AddPlayerBulletPos(bullet, new Vector2(this.Pos.X - 10, this.Pos.Y + 1));
                bullet.SetTask(task);
            }
        }

        private void Shoot2WayRightBullet()
        {
            var task = BulletRunner.ConvertBulletmlTaskOption(Player.b2wayRightBullet);

            if (this.Timer > 0)
            {
                var bullet = new PlayerBullet();
                Manager.AddPlayerBulletPos(bullet, new Vector2(this.Pos.X + 10, this.Pos.Y + 1));
                bullet.SetTask(task);
            }
        }

        private void ShootHomingBullet()
        {
            var task = BulletRunner.ConvertBulletmlTaskOption(Player.homing);

            if (this.Timer > 60)
            {
                var bullet = new PlayerBullet();
                Manager.AddPlayerBulletPos(bullet, this.Pos);
                bullet.SetTask(task);
            }
        }
    }
}
